---
title: "Final_Assignment"
author: "Aaron Malcolm Singleton"
date: "2025-10-05"
output: 
  html_document:
    toc: true
    fig_width: 7
    fig_height: 7
---


```{r setup, include=FALSE}

knitr::opts_chunk$set(
  warning = FALSE,    # hides warnings
  message = FALSE     # hides package messages
)

# Import the gene_expression.tsv file using base R function
gene_expression <- read.table("https://raw.githubusercontent.com/ghazkha/Assessment4/main/gene_expression.tsv", 
                              header = TRUE, sep = "\t", row.names = 1)
```

```{r}
# Show the first six genes
head(gene_expression)
```

```{r}
# Identify all numeric columns (all samples)
numeric_cols <- sapply(gene_expression, is.numeric)

```

```{r}
# Add a new column for mean expression
gene_expression$mean_expression <- rowMeans(gene_expression[, numeric_cols], na.rm = TRUE)

```

```{r}
# Show the first six genes along with their mean expression
head(gene_expression)
```

```{r}
# Sort by mean expression in descending order and display the top 10 genes
top_10_genes <- gene_expression[order(-gene_expression$mean_expression), ]
# Display top 10 genes
head(top_10_genes, 10)

```

```{r}
# Count the number of genes with mean expression < 10
num_genes_low_expression <- sum(gene_expression$mean_expression < 10)
num_genes_low_expression

```

```{r}
# Log-transform mean expression for visualization
log_mean <- log10(gene_expression$mean_expression + 1)
# Plot histogram
hist(log_mean,
     main = "Histogram of Mean Gene Expression (log10)",
     xlab = "log10(Mean Expression + 1)",
     col = "skyblue",
     border = "white")

```

```{r}
# Import the growth_data.csv file
growth_data <- read.csv("https://raw.githubusercontent.com/ghazkha/Assessment4/main/growth_data.csv", header = TRUE)


```

```{r}
# Show the column names
colnames(growth_data)

```

```{r}
# Calculate growth over 2005-2020
growth_data$growth_2005_2020 <- growth_data$Circumf_2020_cm - growth_data$Circumf_2005_cm

# Calculate mean growth by site
mean_growth <- aggregate(growth_2005_2020 ~ Site, data = growth_data, mean)
mean_growth

```

```{r}
# Create a boxplot of growth by site
boxplot(growth_2005_2020 ~ Site, data = growth_data, 
        main = "Tree Growth 2005-2020 by Site",
        xlab = "Site", ylab = "Growth (cm)",
        col = c("lightgreen", "lightcoral"))

```

```{r}
# Perform a t-test to compare growth between the two sites
t_test_result <- t.test(growth_2005_2020 ~ Site, data = growth_data)
t_test_result

```

```{r}
# Calculate mean and standard deviation of tree circumference at start (2005) and end (2020) for each site

# Using aggregate for means
mean_circumf <- aggregate(cbind(Circumf_2005_cm, Circumf_2020_cm) ~ Site, data = growth_data, mean)
mean_circumf

# Using aggregate for standard deviations
sd_circumf <- aggregate(cbind(Circumf_2005_cm, Circumf_2020_cm) ~ Site, data = growth_data, sd)
sd_circumf
```

```{r}
# Create a boxplot of tree circumference at start and end of study for both sites

# Combine data into long format for easier plotting
library(reshape2)
growth_long <- melt(growth_data, id.vars = "Site", measure.vars = c("Circumf_2005_cm", "Circumf_2020_cm"),
                    variable.name = "Year", value.name = "Circumference")

# Boxplot
boxplot(Circumference ~ Site + Year, data = growth_long,
        main = "Tree Circumference at Start and End by Site",
        xlab = "Site and Year", ylab = "Circumference (cm)",
        col = c("lightblue", "lightgreen"),
        names = c("Control 2005", "Control 2020", "Treatment 2005", "Treatment 2020"))
```

```{r}
# Calculate mean growth over last 10 years (2010-2020) at each site

growth_data$growth_2010_2020 <- growth_data$Circumf_2020_cm - growth_data$Circumf_2010_cm
mean_growth_10yr <- aggregate(growth_2010_2020 ~ Site, data = growth_data, mean)
mean_growth_10yr
```

```{r}
# Perform t-test for 10-year growth between sites

t_test_10yr <- t.test(growth_2010_2020 ~ Site, data = growth_data)
t_test_10yr
```

```{r}
#  Visualize 10-year growth distribution by site to make report more informative
boxplot(growth_2010_2020 ~ Site, data = growth_data,
        main = "Tree Growth 2010-2020 by Site",
        xlab = "Site", ylab = "Growth (cm)",
        col = c("lightcoral", "lightgreen"))
```

```{r}
# -------------------------------
# Part 2: 
# Load required libraries
# -------------------------------
library(Biostrings)  # For reading and analyzing DNA and protein sequences
library(ggplot2)
library(dplyr)
library(seqinr)      # For codon usage and k-mer analysis
library(tidyr)       # For data manipulation in plotting codon usage

```

```{r}
# -------------------------------
# Download CDS files
# -------------------------------

# Mycoplasma hyopneumoniae ES-2 CDS (DNA)
mhy_url <- "https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_40_collection/mesomycoplasma_hyopneumoniae_gca_004768725/cds/Mesomycoplasma_hyopneumoniae_gca_004768725.ASM476872v1.cds.all.fa.gz"
mhy_file <- "Mycoplasma_hyopneumoniae_ES2_CDS.fasta"
download.file(mhy_url, mhy_file)

# E. coli K12 CDS (DNA)
ecoli_url <- "https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_30_collection/escherichia_coli_k_12_gca_004803305/cds/Escherichia_coli_k_12_gca_004803305.ASM480330v1.cds.all.fa.gz"
ecoli_gz <- "E_coli_K12_CDS.fa.gz"
ecoli_file <- "E_coli_K12_CDS.fasta"

download.file(ecoli_url, ecoli_gz, mode = "wb")
R.utils::gunzip(ecoli_gz, destname = ecoli_file, overwrite = TRUE)

# --- Load sequences ---
ecoli_cds <- readDNAStringSet(ecoli_file)
org_cds <- readDNAStringSet(mhy_file)

```

```{r}
# -------------------------------
# Count number of CDS
# -------------------------------
num_cds <- data.frame(
  Organism = c("E_coli", "M_hyopneumoniae_ES2"),
  Number_of_CDS = c(length(ecoli_cds), length(org_cds))
)
num_cds

```

```{r}
# -------------------------------
# Total coding DNA length
# -------------------------------
total_coding_length <- data.frame(
  Organism = c("E_coli", "M_hyopneumoniae_ES2"),
  Total_Coding_DNA_bp = c(sum(width(ecoli_cds)), sum(width(org_cds)))
)
total_coding_length

```

```{r}
# -------------------------------
# CDS length distribution
# -------------------------------
cds_lengths <- data.frame(
  Length = c(width(ecoli_cds), width(org_cds)),
  Organism = rep(c("E_coli", "M_hyopneumoniae_ES2"), 
                 times = c(length(ecoli_cds), length(org_cds)))
)

ggplot(cds_lengths, aes(x = Organism, y = Length, fill = Organism)) +
  geom_boxplot() +
  theme_minimal() +
  labs(title = "Coding Sequence Length Distribution", y = "Length (bp)")

# Mean and median CDS length
cds_summary <- cds_lengths %>%
  group_by(Organism) %>%
  summarise(
    Mean_Length = mean(Length),
    Median_Length = median(Length)
  )
cds_summary

```

```{r}
# -------------------------------
# Nucleotide frequency (fixed for single sequence)
# -------------------------------
ecoli_mat <- alphabetFrequency(ecoli_cds, as.prob = TRUE)[, c("A","C","G","T")]
if (is.null(dim(ecoli_mat))) ecoli_mat <- matrix(ecoli_mat, nrow=1)
ecoli_nt_freq <- colMeans(ecoli_mat)

org_mat <- alphabetFrequency(org_cds, as.prob = TRUE)[, c("A","C","G","T")]
if (is.null(dim(org_mat))) org_mat <- matrix(org_mat, nrow=1)
org_nt_freq <- colMeans(org_mat)

nt_freq <- data.frame(
  Base = rep(c("A","C","G","T"), 2),
  Frequency = c(ecoli_nt_freq, org_nt_freq),
  Organism = rep(c("E_coli","M_hyopneumoniae_ES2"), each=4)
)

ggplot(nt_freq, aes(x=Base, y=Frequency, fill=Organism)) +
  geom_bar(stat="identity", position="dodge") +
  labs(title="Nucleotide Frequency", y="Frequency")

```

```{r}
# -------------------------------
# Translate CDS to proteins using Biostrings
# -------------------------------
ecoli_prot <- Biostrings::translate(ecoli_cds)
org_prot <- Biostrings::translate(org_cds)

# Function to calculate amino acid frequency
aa_freq <- function(prot_set) {
  aa_vec <- unlist(strsplit(as.character(prot_set), split=""))
  table(factor(aa_vec, levels=AA_STANDARD)) / length(aa_vec)
}

ecoli_aa_freq <- aa_freq(ecoli_prot)
org_aa_freq <- aa_freq(org_prot)

aa_df <- data.frame(
  AminoAcid = rep(names(ecoli_aa_freq), 2),
  Frequency = c(as.numeric(ecoli_aa_freq), as.numeric(org_aa_freq)),
  Organism = rep(c("E_coli", "M_hyopneumoniae_ES2"), each=20)
)

ggplot(aa_df, aes(x=AminoAcid, y=Frequency, fill=Organism)) +
  geom_bar(stat="identity", position="dodge") +
  labs(title="Amino Acid Frequency", y="Frequency")

```

```{r}
# -------------------------------
# Codon usage 
# -------------------------------

# Function to get codon usage frequencies correctly
get_codon_usage <- function(cds) {
  codons <- unlist(lapply(as.character(cds), function(seq) {
    seq <- gsub("[^ACGT]", "", seq)  # remove invalid chars
    substring(seq, seq(1, nchar(seq) - 2, by = 3), seq(3, nchar(seq), by = 3))
  }))
  table(codons)
}

# Calculate codon usage for each organism
ecoli_codon_usage <- get_codon_usage(ecoli_cds)
org_codon_usage <- get_codon_usage(org_cds)

# Combine into one data frame
codon_df <- data.frame(
  Codon = union(names(ecoli_codon_usage), names(org_codon_usage))
) %>%
  mutate(
    E_coli = ecoli_codon_usage[Codon],
    M_hyopneumoniae_ES2 = org_codon_usage[Codon]
  ) %>%
  replace(is.na(.), 0)

# Take the first 20 codons for example plotting
codon_melt <- codon_df[1:20, ] %>%
  tidyr::pivot_longer(cols = -Codon, names_to = "Organism", values_to = "Frequency")

# Plot
ggplot(codon_melt, aes(x = Codon, y = Frequency, fill = Organism)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  labs(title = "Codon Usage Frequency (First 20 Codons)", y = "Frequency") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

```{r}
# -------------------------------
# K-mer analysis for proteins (length 3)
# -------------------------------
count_kmers <- function(prot_set, k) {
  aa_seq <- unlist(strsplit(as.character(prot_set), ""))
  kmers <- sapply(1:(length(aa_seq)-k+1), function(i) paste(aa_seq[i:(i+k-1)], collapse=""))
  sort(table(kmers), decreasing=TRUE)
}

# Top 10 overrepresented 3-mers in Mycoplasma
org_kmers_3 <- count_kmers(org_prot, 3)
head(org_kmers_3, 10)

# Compare with E. coli
ecoli_kmers_3 <- count_kmers(ecoli_prot, 3)
head(ecoli_kmers_3, 10)


# Prepare data: take top 10 k-mers for each organism
top_org_kmers <- head(org_kmers_3, 10)
top_ecoli_kmers <- head(ecoli_kmers_3, 10)

# Combine into a data frame
kmers_df <- data.frame(
  Kmer = c(names(top_org_kmers), names(top_ecoli_kmers)),
  Frequency = c(as.numeric(top_org_kmers), as.numeric(top_ecoli_kmers)),
  Organism = rep(c("M. hyopneumoniae ES2", "E. coli"), each = 10)
)

# Plot
ggplot(kmers_df, aes(x = reorder(Kmer, -Frequency), y = Frequency, fill = Organism)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "Top 10 Overrepresented 3-mers in Proteins",
       x = "3-mer",
       y = "Frequency") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Bottom 10 (underrepresented) 3-mers for each organism
bottom_org_kmers <- tail(org_kmers_3, 10)
bottom_ecoli_kmers <- tail(ecoli_kmers_3, 10)

# Combine into a data frame
under_kmers_df <- data.frame(
  Kmer = c(names(bottom_org_kmers), names(bottom_ecoli_kmers)),
  Frequency = c(as.numeric(bottom_org_kmers), as.numeric(bottom_ecoli_kmers)),
  Organism = rep(c("M. hyopneumoniae ES2", "E. coli"), each = 10)
)

# Plot
ggplot(under_kmers_df, aes(x = reorder(Kmer, Frequency), y = Frequency, fill = Organism)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "Top 10 Underrepresented 3-mers in Proteins",
       x = "3-mer",
       y = "Frequency") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))



```





