---
title: "Final_Assignment"
author: "Aaron Malcolm Singleton"
date: "2025-10-05"
output: 
  html_document:
    toc: true
    fig_width: 7
    fig_height: 7
---


```{r setup, include=FALSE}

knitr::opts_chunk$set(
  warning = FALSE,    # hides warnings
  message = FALSE     # hides package messages
)

# Import the gene_expression.tsv file using base R function
gene_expression <- read.table("https://raw.githubusercontent.com/ghazkha/Assessment4/main/gene_expression.tsv", 
                              header = TRUE, sep = "\t", row.names = 1)
```

```{r}
# Question 1: Show the first six genes
head(gene_expression)

#Description:
#I imported a tab-separated values file containing gene expression data. Checking the first rows ensures the data loaded correctly.

```

```{r}
# Identify all numeric columns (all samples)
numeric_cols <- sapply(gene_expression, is.numeric)
#Description:
#This selects columns with numeric values, which represent gene expression measurements. This step ensures that calculations like mean expression are only done on numeric data.
```

```{r}
# Question 2: Add a new column for mean expression
gene_expression$mean_expression <- rowMeans(gene_expression[, numeric_cols], na.rm = TRUE)

```

```{r}
# Show the first six genes along with their mean expression
head(gene_expression)
#Description:
#This selects columns with numeric values, which represent gene expression measurements. This step ensures that calculations like mean expression are only done on numeric data. I compute the average expression for each gene across all samples. na.rm = TRUE ensures missing values do not affect the calculation.
```

```{r}
# Question 3: Sort by mean expression in descending order and display the top 10 genes
top_10_genes <- gene_expression[order(-gene_expression$mean_expression), ]
# Display top 10 genes
head(top_10_genes, 10)
#Description:
#I ranked genes by their mean expression to find the 10 most highly expressed genes.

```

```{r}
# Question 4: Count the number of genes with mean expression < 10
num_genes_low_expression <- sum(gene_expression$mean_expression < 10)
num_genes_low_expression
#Description:
#This identifies genes with very low expression, which might be considered inactive or not expressed under the given conditions.
```

```{r}
# Question 5: Log-transform mean expression for visualization
log_mean <- log10(gene_expression$mean_expression + 1)
# Plot histogram
hist(log_mean,
     main = "Histogram of Mean Gene Expression (log10)",
     xlab = "log10(Mean Expression + 1)",
     col = "skyblue",
     border = "white")

#Description:
#A histogram of log-transformed mean expression helps visualize the distribution of gene expression, especially when values vary widely.
```

```{r}
# Question 6: Import the growth_data.csv file
growth_data <- read.csv("https://raw.githubusercontent.com/ghazkha/Assessment4/main/growth_data.csv", header = TRUE)

```

```{r}
# Show the column names
colnames(growth_data)

#Description:
#I loaded a dataset containing tree circumference measurements for different sites and years. Checking column names ensures proper referencing in subsequent calculations.
```

```{r}
# Calculate growth over 2005-2020
growth_data$growth_2005_2020 <- growth_data$Circumf_2020_cm - growth_data$Circumf_2005_cm

# Calculate mean growth by site
mean_growth <- aggregate(growth_2005_2020 ~ Site, data = growth_data, mean)
mean_growth

#Description:
#Growth is measured as the difference in circumference over 15 years. We then calculate the average growth for each site.
```

```{r}
# Create a boxplot of growth by site
boxplot(growth_2005_2020 ~ Site, data = growth_data, 
        main = "Tree Growth 2005-2020 by Site",
        xlab = "Site", ylab = "Growth (cm)",
        col = c("lightgreen", "lightcoral"))
#Description:
#Growth is measured as the difference in circumference over 15 years. We then calculate the average growth for each site.

```

```{r}
# Perform a t-test to compare growth between the two sites
t_test_result <- t.test(growth_2005_2020 ~ Site, data = growth_data)
t_test_result

#Description:
#Growth is measured as the difference in circumference over 15 years. We then calculate the average growth for each site.
```

```{r}
# Question 7: Calculate mean and standard deviation of tree circumference at start (2005) and end (2020) for each site

# Using aggregate for means
mean_circumf <- aggregate(cbind(Circumf_2005_cm, Circumf_2020_cm) ~ Site, data = growth_data, mean)
mean_circumf

# Using aggregate for standard deviations
sd_circumf <- aggregate(cbind(Circumf_2005_cm, Circumf_2020_cm) ~ Site, data = growth_data, sd)
sd_circumf

#Description:
#I summarize tree size at the start and end of the study, including variability (SD).

```

```{r}
# Create a boxplot of tree circumference at start and end of study for both sites

# Combine data into long format for easier plotting
library(reshape2)
growth_long <- melt(growth_data, id.vars = "Site", measure.vars = c("Circumf_2005_cm", "Circumf_2020_cm"),
                    variable.name = "Year", value.name = "Circumference")

# Boxplot
boxplot(Circumference ~ Site + Year, data = growth_long,
        main = "Tree Circumference at Start and End by Site",
        xlab = "Site and Year", ylab = "Circumference (cm)",
        col = c("lightblue", "lightgreen"),
        names = c("Control 2005", "Control 2020", "Treatment 2005", "Treatment 2020"))
#Description:
#We reshape the data to long format and create a boxplot to visualize tree sizes at both years for each site.

```

```{r}
# Question 9: Calculate mean growth over last 10 years (2010-2020) at each site

growth_data$growth_2010_2020 <- growth_data$Circumf_2020_cm - growth_data$Circumf_2010_cm
mean_growth_10yr <- aggregate(growth_2010_2020 ~ Site, data = growth_data, mean)
mean_growth_10yr
```

```{r}
# Question 10: Perform t-test for 10-year growth between sites

t_test_10yr <- t.test(growth_2010_2020 ~ Site, data = growth_data)
t_test_10yr
```

```{r}
#  Visualize 10-year growth distribution by site to make report more informative
boxplot(growth_2010_2020 ~ Site, data = growth_data,
        main = "Tree Growth 2010-2020 by Site",
        xlab = "Site", ylab = "Growth (cm)",
        col = c("lightcoral", "lightgreen"))
```

```{r}
# -------------------------------
# Part 2: 
# Load required libraries
# -------------------------------
library(seqinr)
library(R.utils)

#Description 
# This chunk loads essential R libraries. 
# 'seqinr' is used for sequence reading, translation, and codon analysis. 
# 'R.utils' is used to decompress .gz files.


```

```{r}
# -------------------------------
# Download CDS files
# -------------------------------

# Mycoplasma hyopneumoniae ES-2 CDS
mhy_url <- "https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_40_collection/mesomycoplasma_hyopneumoniae_gca_004768725/cds/Mesomycoplasma_hyopneumoniae_gca_004768725.ASM476872v1.cds.all.fa.gz"
mhy_file <- "M_hyopneumoniae_ES2_CDS.fasta"
download.file(mhy_url, destfile = "M_hyopneumoniae_ES2_CDS.fa.gz", mode="wb")
R.utils::gunzip("M_hyopneumoniae_ES2_CDS.fa.gz", destname = mhy_file, overwrite = TRUE)

# E. coli K12 CDS
ecoli_url <- "https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_30_collection/escherichia_coli_k_12_gca_004803305/cds/Escherichia_coli_k_12_gca_004803305.ASM480330v1.cds.all.fa.gz"
ecoli_file <- "E_coli_K12_CDS.fasta"
download.file(ecoli_url, destfile = "E_coli_K12_CDS.fa.gz", mode="wb")
R.utils::gunzip("E_coli_K12_CDS.fa.gz", destname = ecoli_file, overwrite = TRUE)

# --- Load sequences using seqinr ---
org_cds <- read.fasta(file = mhy_file, seqtype = "DNA", as.string = TRUE)
ecoli_cds <- read.fasta(file = ecoli_file, seqtype = "DNA", as.string = TRUE)

#Description:
# Downloads the CDS (coding sequences) of *E. coli* and *M. hyopneumoniae* from Ensembl Bacteria.
# Files are uncompressed with gunzip() and read into R using read.fasta().


```

```{r}
# -------------------------------
# Question 1: Count number of CDS
# -------------------------------
num_cds <- data.frame(
  Organism = c("E_coli", "M_hyopneumoniae_ES2"),
  Number_of_CDS = c(length(ecoli_cds), length(org_cds))
)
knitr::kable(num_cds, caption="Number of CDS in each organism")

#Description:
# Calculates the number of CDS in each organism. 
# *E. coli* has more CDS than *M. hyopneumoniae*, reflecting its larger genome and gene content.


```

```{r}
# -------------------------------
# Question 2: Total coding DNA length
# -------------------------------
len_ecoli <- sapply(ecoli_cds, nchar)
len_org   <- sapply(org_cds, nchar)

total_coding_length <- data.frame(
  Organism = c("E_coli", "M_hyopneumoniae_ES2"),
  Total_Coding_DNA_bp = c(sum(len_ecoli), sum(len_org)),
  Mean_CDS_Length = c(mean(len_ecoli), mean(len_org)),
  Median_CDS_Length = c(median(len_ecoli), median(len_org))
)
knitr::kable(total_coding_length, caption="Total coding DNA and CDS length summary")

# Boxplot of CDS lengths
boxplot(list(E_coli=len_ecoli, M_hyopneumoniae_ES2=len_org),
        ylab="CDS length (bp)", main="CDS length distributions")

#Description:
# Calculates total, mean, and median CDS lengths for each organism. 
# *E. coli* has more total coding DNA, while *M. hyopneumoniae* has longer average CDS, 
# likely due to its compact genome and fewer genes.


```

```{r}
# -------------------------------
# Question 3: DNA base frequencies
# -------------------------------
dna_ecoli <- unlist(lapply(ecoli_cds, function(seq) s2c(seq[[1]])))
dna_org   <- unlist(lapply(org_cds,   function(seq) s2c(seq[[1]])))

# Base frequencies
basefreq_ecoli <- count(dna_ecoli, 1)
basefreq_org   <- count(dna_org, 1)

# Plot DNA base frequencies
barplot(as.numeric(basefreq_ecoli), names.arg=names(basefreq_ecoli),
        main="E_coli DNA base frequency", ylab="Count")
barplot(as.numeric(basefreq_org), names.arg=names(basefreq_org),
        main="M_hyopneumoniae_ES2 DNA base frequency", ylab="Count")
#Description:
# Calculates base composition (A, T, C, G) for CDS in each organism.
# *M. hyopneumoniae* shows AT-rich content, typical of Mycoplasmas, 
# whereas *E. coli* has a more balanced base composition.



```

```{r}
# -------------------------------
# Question 4: Translate CDS to protein and AA frequencies
# -------------------------------

# Translate CDS into protein
prot_ecoli <- lapply(ecoli_cds, function(seq) translate(s2c(seq[[1]])))
prot_org   <- lapply(org_cds,   function(seq) translate(s2c(seq[[1]])))

# Amino acid frequencies
aa <- unique(unlist(prot_ecoli))
aa <- aa[aa != "*"]

aafreq_ecoli <- count(unlist(prot_ecoli), wordsize=1, alphabet=aa)
aafreq_org   <- count(unlist(prot_org),   wordsize=1, alphabet=aa)

barplot(as.numeric(aafreq_ecoli), names.arg=names(aafreq_ecoli),
        las=2, main="E_coli amino acid frequencies", ylab="Count")
barplot(as.numeric(aafreq_org), names.arg=names(aafreq_org),
        las=2, main="M_hyopneumoniae_ES2 amino acid frequencies", ylab="Count")


#Description:
# Translates CDS into protein sequences and calculates amino acid frequencies.
# Differences in amino acid usage reflect codon preferences, genome composition, and protein requirements.



```

```{r}
# -------------------------------
# Question 5: Codon usage bias (RSCU)
# -------------------------------
# Codon usage (RSCU) example
rscu_ecoli <- uco(dna_ecoli, index="rscu", as.data.frame=TRUE)
rscu_org   <- uco(dna_org,   index="rscu", as.data.frame=TRUE)

barplot(rscu_ecoli$freq, names.arg=rscu_ecoli$codon,
        las=2, main="E_coli codon usage (RSCU)", cex.names=0.6, ylab="RSCU")
barplot(rscu_org$freq, names.arg=rscu_org$codon,
        las=2, main="M_hyopneumoniae_ES2 codon usage (RSCU)", cex.names=0.6, ylab="RSCU")

## Computes and plots Relative Synonymous Codon Usage (RSCU) values.
# *M. hyopneumoniae* shows strong codon bias due to its AT-rich genome.
# *E. coli* displays more uniform codon usage, reflecting broader codon preference.


```

```{r}
# -------------------------------
# Q6: Protein k-mers (3-5)
# -------------------------------

# Convert CDS strings to character vectors and translate
prots_ecoli <- unlist(lapply(ecoli_cds, function(seq) translate(s2c(seq[[1]]))))
prots_org   <- unlist(lapply(org_cds,   function(seq) translate(s2c(seq[[1]]))))


# Amino acid alphabet (drop stop codon '*')
aa <- unique(prots_ecoli)
aa <- aa[aa != "*"]

# k = 3
k3_ecoli <- count(prots_ecoli, wordsize=3, alphabet=aa)
k3_org <- count(prots_org, wordsize=3, alphabet=aa)

top_over_org3  <- head(sort(k3_org, decreasing=TRUE), 10)
top_under_org3 <- head(sort(k3_org, decreasing=FALSE), 10)

barplot(as.numeric(top_over_org3), names.arg=names(top_over_org3),
        las=2, main="M_hyopneumoniae_ES2: Top 10 over-represented 3-mers")
barplot(as.numeric(top_under_org3), names.arg=names(top_under_org3),
        las=2, main="M_hyopneumoniae_ES2: Top 10 under-represented 3-mers")

# Compare with E. coli counts for same k-mers
k3_ecoli[names(top_over_org3)]
k3_ecoli[names(top_under_org3)]

# k = 4
k4_org <- count(prots_org, wordsize=4, alphabet=aa)
top_over_org4  <- head(sort(k4_org, decreasing=TRUE), 10)
top_under_org4 <- head(sort(k4_org, decreasing=FALSE),10)

# k = 5
k5_org <- count(prots_org, wordsize=5, alphabet=aa)
top_over_org5  <- head(sort(k5_org, decreasing=TRUE), 10)
top_under_org5 <- head(sort(k5_org, decreasing=FALSE),10)


#Description:
## Calculates 3–5-mer amino acid patterns in protein sequences.
# Plots the most and least frequent 3-mers in *M. hyopneumoniae* and compares them with *E. coli*.
# These patterns reflect differences in codon usage, protein structure, and genome evolution.

```





